### PR TITLE
1626 | SNC 2.0.0 | Update SNController to support SDK/FW v3.0.0

### PR DESCRIPTION
**Pull request by Salonso1602**
# Resolves  
https://focusuy.atlassian.net/browse/BMC2-1626 

# How to test  
Run `pip install .`
Run tests  
`python tests/test.py`
`python tests/uart_tests.py`

# What to expect
All tests using `USE_REAL_DEVICE=True` should pass

# Test Rationale and Changes  
What was updated here was how the errors are parsed in the "init bus"-esque methods in the different protocols (such as set_bus_voltage, set_pin_voltage, etc) and particularly how the success message looks for I3C
The tests utilize these init bus methods

### CODE REVIEW

**Comment by pons93** [comment]
Result for `test.py`:
```
s.s................Fssssssss.s..s.....s...
======================================================================
FAIL: test_i3c_init_bus_with_no_targets_connected (__main__.TestSupernovaController)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\test.py", line 198, in test_i3c_init_bus_with_no_targets_connected
    self.assertEqual(success, False)
AssertionError: True != False

----------------------------------------------------------------------
Ran 42 tests in 9.579s

FAILED (failures=1, skipped=13)
```

**Comment by Salonso1602** [comment]
@pons93 The result youve presented for the uart_tests is expected, since those tests fail due to lack of error handling in UART which is fixed @ #69

**Comment by pons93** [comment]
> @pons93 The result youve presented for the uart_tests is expected, since those tests fail due to lack of error handling in UART which is fixed @ #69

Ok. And I guess the difference in the i3c test error is because the new firmwares 2.5.0 and 3.0.0 have removed the errors field from the response structure.

**Review by pons93 — CHANGES_REQUESTED** [review]
When I try to test with fw 2.5.0 with no I3C target connected, the tests fails do to `KeyErrors` in the `i3c.py` file:

```
sEsE..EE...........Essssssss.s..s.....s..E
======================================================================
ERROR: test_broadcast_rstact (__main__.TestSupernovaController)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\test.py", line 552, in test_broadcast_rstact
    i3c.init_bus(3300)
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\i3c.py", line 192, in init_bus
    result = (False, {"errors": responses[0]["errors"]})
KeyError: 'errors'

======================================================================
ERROR: test_ccc_rstdaa (__main__.TestSupernovaController)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\test.py", line 369, in test_ccc_rstdaa
    i3c.init_bus(3300)
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\i3c.py", line 192, in init_bus
    result = (False, {"errors": responses[0]["errors"]})
KeyError: 'errors'

======================================================================
ERROR: test_direct_rstact_read (__main__.TestSupernovaController)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\test.py", line 528, in test_direct_rstact_read
    i3c.init_bus(3300)
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\i3c.py", line 192, in init_bus
    result = (False, {"errors": responses[0]["errors"]})
KeyError: 'errors'

======================================================================
ERROR: test_direct_rstact_write (__main__.TestSupernovaController)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\test.py", line 540, in test_direct_rstact_write
    i3c.init_bus(3300)
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\i3c.py", line 192, in init_bus
    result = (False, {"errors": responses[0]["errors"]})
KeyError: 'errors'

======================================================================
ERROR: test_i3c_init_bus_with_no_targets_connected (__main__.TestSupernovaController)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\test.py", line 196, in test_i3c_init_bus_with_no_targets_connected
    (success, result) = i3c.init_bus(1500)
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\i3c.py", line 192, in init_bus
    result = (False, {"errors": responses[0]["errors"]})
KeyError: 'errors'

======================================================================
ERROR: test_trigger_target_reset_pattern (__main__.TestSupernovaController)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\test.py", line 564, in test_trigger_target_reset_pattern
    i3c.init_bus(3300)
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\i3c.py", line 192, in init_bus
    result = (False, {"errors": responses[0]["errors"]})
KeyError: 'errors'

----------------------------------------------------------------------
Ran 42 tests in 12.404s

FAILED (errors=6, skipped=13)
```

How can we test and be sure that in failing cases, we are not trying to access to non-existent keys?

**Comment by Salonso1602** [comment]
whilst tests could fail depending on certain hardware configurations, errors on tests should never happen
A fix to what you mentioned has been pushed

**Comment by Salonso1602** [comment]
I am passing this PR to draft to focus on supporting FW/SDK 3.0.0 directly as we have spoken

**Review by pons93 — COMMENTED** [review]
Using fw 3.0.0 I obtained that all tests in test.py passed.

However, in uart_tests.py:
```
..FF..
======================================================================
FAIL: test_uart_init_bus (__main__.TestSupernovaControllerUART)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\uart_tests.py", line 94, in test_uart_init_bus
    self.assertEqual(success, True, f"Failed: {response}")
AssertionError: False != True : Failed: Init failed, error from the Supernova

======================================================================
FAIL: test_uart_multiple_init_bus (__main__.TestSupernovaControllerUART)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\uart_tests.py", line 108, in test_uart_multiple_init_bus
    self.assertEqual(True, success, f"Failed on first init: {response}")
AssertionError: True != False : Failed on first init: Init failed, error from the Supernova

----------------------------------------------------------------------
Ran 6 tests in 7.876s

FAILED (failures=2)
```

and in i2c tests:
```
......EE..
======================================================================
ERROR: test_i2c_write_non_stop (__main__.TestSupernovaController)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\i2c_tests.py", line 68, in test_i2c_write_non_stop
    (success, result) = self.i2c.write_non_stop(0x50, [0x01,0x00], data_to_write)
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\i2c.py", line 294, in write_non_stop
    result = (False, CodeTranslator.get_message("i2c", responses[0]["status"]))
NameError: name 'CodeTranslator' is not defined

======================================================================
ERROR: test_i2c_write_non_stop_should_nack (__main__.TestSupernovaController)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\i2c_tests.py", line 85, in test_i2c_write_non_stop_should_nack
    (success, result) = self.i2c.write_non_stop(0x56, [0x01,0x00], data_to_write)
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\i2c.py", line 294, in write_non_stop
    result = (False, CodeTranslator.get_message("i2c", responses[0]["status"]))
NameError: name 'CodeTranslator' is not defined

----------------------------------------------------------------------
Ran 10 tests in 5.469s

FAILED (errors=2)
```

**Comment by Salonso1602** [comment]
The Uart errors are expected (see the in code comments
The i2c errors seem to relate to write non stop, which was added in 1.5.0. maybe a rebase went awry, i ll investigate it

File: supernovacontroller/sequential/i3c.py [hunk 1]
```diff
@@ -185,10 +186,10 @@ def init_bus(self, voltage: int=None, targets=None):
         # TODO: Toggle IBIs off
 
         status = responses[0]["result"]
-        if status == "I3C_BUS_INIT_SUCCESS":
+        if status == "DAA_SUCCESS":
```
Comment (pons93) [inline supernovacontroller/sequential/i3c.py:189]: Shouldn't we also check that response["descriptor"]["errors"] = "NO_TRANSFER_ERROR" ?

File: supernovacontroller/sequential/i3c.py [hunk 2]
```diff
@@ -185,10 +186,10 @@ def init_bus(self, voltage: int=None, targets=None):
...
             result = (True, voltage)
         else:
-            result = (False, {"errors": responses[0]["errors"]})
+            result = (False, {"errors": responses[0]["result"]})
```
Comment (pons93) [inline supernovacontroller/sequential/i3c.py:192]: This contains also if comes something in `invalidAddresses`?

**Review by pons93 — CHANGES_REQUESTED** [review]

File: supernovacontroller/sequential/i3c.py (hunk 2 above)
Comment (Salonso1602) [reply supernovacontroller/sequential/i3c.py:192]: Is it possible to get invalid addresses via the init bus command?

**Review by pons93 — APPROVED** [review]
LGTM!!!!!!

Great job!!!

**Review by joaquin-urrisa — APPROVED** [review]
LGTM!

**Pull request merged**